openvdb/tools/FastSweeping.h

```diff
@@ -0,0 +1,1432 @@
+#include <openvdb/tree/LeafManager.h>
+#include "LevelSetUtil.h"
+#include "Morphology.h"
+
+#include "Statistics.h"
+#ifdef BENCHMARK_FAST_SWEEPING
+#include <openvdb/util/CpuTimer.h>
+#endif
+
+#include <openvdb/math/Stats.h>
+
+namespace openvdb {
+OPENVDB_USE_VERSION_NAMESPACE
+namespace OPENVDB_VERSION_NAME {
+namespace tools {
+
+/// @brief Converts a scalar fog volume into a signed distance function. Active input voxels
+///        with scalar values above the given isoValue will have NEGATIVE distance
+///        values on output, i.e. they are assumed to be INSIDE the iso-surface.
+///
+/// @return A shared pointer to a signed-distance field defined on the active values
+///         of the input fog volume.
+///
+/// @param fogGrid  Scalar (floating-point) volume from which an
+///                 iso-surface can be defined.
+///
+/// @param isoValue A value which defines a smooth iso-surface that
+///                 intersects active voxels in @a fogGrid.
+///
+/// @param nIter    Number of iterations of the fast sweeping algorithm.
+///                 each performing 2^3 = 8 sweeps.
+///
+/// @note Strictly speaking a fog volume is normalized to the range [0,1] but this
+///       method accepts a scalar volume with an arbitary range, as long as the it
+///       includes the @a isoValue.
+///
+/// @details Topology of output grid is identical to that of the input grid, except
+///          active tiles in the input grid will be converted to active voxels
+///          in the output grid!
+///
+/// @throw  RuntimeError if the iso-surface does not intersected any active
+///         voxels or if it intersects any active tiles in @a fogGrid.
+template<typename GridT>
+typename GridT::Ptr
+fogToSdf(const GridT &fogGrid,
+         typename GridT::ValueType isoValue,
+         int nIter = 1);
+
+/// @brief Given an existing approximate SDF it solves the Eikonal equation for all its
+///        active voxels. Active input voxels with a signed distance value above the
+///        given isoValue will have POSITIVE distance values on output, i.e. they are
+///        assumed to be OUTSIDE the iso-surface.
+///
+/// @return A shared pointer to a signed-distance field defined on the active values
+///         of the input sdf volume.
+///
+/// @param sdfGrid  An approximate signed distance field to the specified iso-surface.
+///
+/// @param isoValue A value which defines a smooth iso-surface that
+///                 intersects active voxels in @a sdfGrid.
+///
+/// @param nIter    Number of iterations of the fast sweeping algorithm.
+///                 each performing 2^3 = 8 sweeps.
+///
+/// @note The only difference between this method and fogToSdf, defined above, is the
+///       convention of the sign of the output distance field.
+///
+/// @details Topology of output grid is identical to that of the input grid, except
+///          active tiles in the input grid will be converted to active voxels
+///          in the output grid!
```

Comment:
Isn't that implied because we are building an SDF, so the values must vary, so naturally constant tiles will densify?

Reply:
sure, but it doesn't hurt to explain it does it?

Reply:
:>  Yeah, I think this is my bias to want transparent constant tiles :>